test/system/src/main/java/io/pravega/test/system/framework/services/docker/DockerBasedService.java

```diff
@@ -0,0 +1,158 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+@Slf4j
+public abstract class DockerBasedService  implements io.pravega.test.system.framework.services.Service {
+
+    static final int ZKSERVICE_ZKPORT = 2181;
+    static final int CONTROLLER_PORT = 9092;
+    static final String IMAGE_PATH = System.getProperty("dockerImageRegistry");
+    static final String PRAVEGA_VERSION = System.getProperty("imageVersion");
+    DockerClient dockerClient;
+    String serviceName;
+    private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(3);
+
+    DockerBasedService(String serviceName) {
+        dockerClient = DefaultDockerClient.builder().uri("http://"+ System.getProperty("masterIP")+ ":2375").build();
+        try {
+            if (dockerClient.listNetworks(DockerClient.ListNetworksParam.byNetworkName("docker-network")).isEmpty()) {
+                dockerClient.createNetwork(NetworkConfig.builder().driver("overlay").name("docker-network").build());
+            }
+        } catch (DockerException | InterruptedException e) {
+            log.error("Unable to create network {}", e);
+        }
+        this.serviceName = serviceName;
+        }
+
+
+    @Override
+    public String getID() {
+        Service.Criteria criteria = Service.Criteria.builder().serviceName(this.serviceName).build();
+        String serviceId = null;
+        try {
+            List<Service> serviceList = dockerClient.listServices(criteria);
+            serviceId = serviceList.get(0).id();
+        }  catch (DockerException | InterruptedException  e) {
+            log.error("Unable to get service id {}", e);
+        }
+        return serviceId;
+    }
+
+    @Override
+    public boolean isRunning() {
+        boolean value = false;
+        try {
+            //list the service with filter 'serviceName'
+            Service.Criteria criteria = Service.Criteria.builder().serviceName(this.serviceName).build();
+            List<Service> serviceList = dockerClient.listServices(criteria);
+
+            for (int i = 0; i < serviceList.size(); i++) {
+                String serviceId = serviceList.get(i).id();
+                if (!serviceId.isEmpty()) {
+                    value = true;
+                    break;
+                }
+            }
+        } catch (DockerException | InterruptedException e) {
+            log.error("Unable to list docker services {}", e);
+        }
+        return value;
+    }
+
+    CompletableFuture<Void> waitUntilServiceRunning() {
+        log.debug("IS RUNNING {}", isRunning());
+        return FutureHelpers.loop(() -> !isRunning(), //condition
+                () -> FutureHelpers.delayedFuture(Duration.ofSeconds(5), executorService),
+                executorService);
+    }
+
+    String setSystemProperty(final String propertyName, final String propertyValue) {
```

Comment:
This isn't a great abstraction because it just causes the callers to call it over and over again concating the results. Perhaps it should take a string builder. Also this method could be static.

Reply:
done.